crates/ruff_cli/src/printer.rs

```diff
@@ -471,41 +471,40 @@ impl Printer {
     }
 
     pub fn write_statistics(&self, diagnostics: &Diagnostics) -> Result<()> {
-        let violations: Vec<Rule> = diagnostics
+        let statistics: Vec<ExpandedStatistics> = diagnostics
             .messages
             .iter()
-            .map(|message| message.kind.rule())
+            .map(|message| {
+                (
+                    message.kind.rule().noqa_code(),
+                    message.kind.body.clone(),
```

Comment:
Is it necessary to clone `body` here, and later on too? Not a huge deal, but it seems wasteful given that we only end up using the first `body` anyway.

Reply:
I tried sorting and iterating over `Message` instead, but it is slower by more than 2x.
You can go ahead and try to optimize if you find something better.

Reply:
All good, maybe I'll play around with the ownership but this is clearly a huge improvement!

Reply:
Do you mind testing the version I just pushed?